File: backend/app/schemas/hoja_vida.py
"""
Schemas para hoja de vida reproductiva del animal.
"""
from datetime import date
from typing import Optional

from pydantic import BaseModel, Field


class AnimalResumenGenealogia(BaseModel):
    id: int
    numero_identificacion: str
    nombre: Optional[str] = None
    sexo: str
    fecha_nacimiento: Optional[date] = None


class EventoHojaVidaReproductiva(BaseModel):
    id: int
    tipo_evento: str
    fecha_evento: date
    diagnostico: Optional[str] = None
    tipo_servicio: Optional[str] = None
    toro_numero: Optional[str] = None
    toro_nombre: Optional[str] = None
    hembra_numero: Optional[str] = None
    hembra_nombre: Optional[str] = None
    numero_crias: Optional[int] = None
    tipo_parto: Optional[str] = None
    facilidad_parto: Optional[str] = None
    vitalidad_cria: Optional[str] = None
    crias_registradas: list[AnimalResumenGenealogia] = Field(default_factory=list)


class HojaVidaReproductivaResponse(BaseModel):
    madre: Optional[AnimalResumenGenealogia] = None
    padre: Optional[AnimalResumenGenealogia] = None
    crias_en_inventario: list[AnimalResumenGenealogia] = Field(default_factory=list)
    progenie_como_padre: list[AnimalResumenGenealogia] = Field(default_factory=list)
    eventos: list[EventoHojaVidaReproductiva] = Field(default_factory=list)

File: frontend/src/types/hojaVida.ts
export interface AnimalResumenGenealogia {
  id: number;
  numero_identificacion: string;
  nombre?: string | null;
  sexo: string;
  fecha_nacimiento?: string | null;
}

export interface EventoHojaVidaReproductiva {
  id: number;
  tipo_evento: string;
  fecha_evento: string;
  diagnostico?: string | null;
  tipo_servicio?: string | null;
  toro_numero?: string | null;
  toro_nombre?: string | null;
  hembra_numero?: string | null;
  hembra_nombre?: string | null;
  numero_crias?: number | null;
  tipo_parto?: string | null;
  facilidad_parto?: string | null;
  vitalidad_cria?: string | null;
  crias_registradas: AnimalResumenGenealogia[];
}

export interface HojaVidaReproductiva {
  madre?: AnimalResumenGenealogia | null;
  padre?: AnimalResumenGenealogia | null;
  crias_en_inventario: AnimalResumenGenealogia[];
  progenie_como_padre: AnimalResumenGenealogia[];
  eventos: EventoHojaVidaReproductiva[];
}

export function formatAnimalResumen(animal: AnimalResumenGenealogia): string {
  const nombre = animal.nombre?.trim();
  if (nombre) {
    return `${animal.numero_identificacion} — ${nombre}`;
  }
  return animal.numero_identificacion;
}

export function labelTipoEventoReproductivo(tipo: string): string {
  const map: Record<string, string> = {
    servicio: 'Servicio / monta',
    diagnostico: 'Diagnóstico',
    parto: 'Parto',
    aborto: 'Aborto',
    secado: 'Secado',
    otro: 'Otro',
  };
  return map[tipo] || tipo;
}
